### ISSUE TITLE
[Design] Failover Logic between Providers and Models

### ISSUE THREAD

**Issue by missBerg** [theagentrouter/agent-router#34]
The design proposal should include:

- [ ] Motivation
- [ ] Feature Definition
- [ ] Control Plane API
- [ ] Technical Implementation Proposal

**This issue was created from conversation during Dec 5th Community meeting**
https://docs.google.com/document/d/10e1sfsF-3G3Du5nBHGmLjXw5GVMqqCvFDqp_O65B0_w/edit?tab=t.0#bookmark=id.dz9gpy397ymu

**Comment by yuzisun** [comment]
This can potentially leverage the implementation in envoy gateway though we will to think how to expose on `AIGatewayRoute`
https://gateway.envoyproxy.io/docs/tasks/traffic/failover/

**Comment by mathetake** [comment]
so what we essentially need is to "go through the extproc filter chain again to perform the necessary mutation" to re-mutate auth header, re-perform transformation, etc and that's not possible by either default Envoy or Inference Extension spec. 

One workaround I can think of is to make a request directly from the extproc to the same listener where the extproc is running when the response header for the initially chosen backend is 5xx. When directly sending a request from the extproc, we can add a special header like #73 to specifically choose the backend without running the router's matching. Then, if the response from the second request originating from the extproc is 200, we can use ImmediateResponse to send the response crafted from the contents from the fallback request to the client.


But this comes with a cost; until we receive the first 200 response, the extproc needs to retain the whole request body (and this is inevitable by the nature of this type of fallback regardless of Envoy or whatever proxy we use).


The good thing about this workaround is that all the logic is implementable without any changes in EG or Envoy themselves

**Comment by yuzisun** [comment]
> so what we essentially need is to "go through the extproc filter chain again to perform the necessary mutation" to re-mutate auth header, re-perform transformation, etc and that's not possible by either default Envoy or Inference Extension spec.
> 
> One workaround I can think of is to make a request directly from the extproc to the same listener where the extproc is running when the response header for the initially chosen backend is 5xx. When directly sending a request from the extproc, we can add a special header like [#73](https://github.com/envoyproxy/ai-gateway/issues/73) to specifically choose the backend without running the router's matching. Then, if the response from the second request originating from the extproc is 200, we can use ImmediateResponse to send the response crafted from the contents from the fallback request to the client.
> 
> But this comes with a cost; until we receive the first 200 response, the extproc needs to retain the whole request body (and this is inevitable by the nature of this type of fallback regardless of Envoy or whatever proxy we use).
> 
> The good thing about this workaround is that all the logic is implementable without any changes in EG or Envoy themselves

@mathetake If we send the retry request directly to the extproc listener from the extproc response handler, how does envoy proxy know to execute the remaining filters to route the retry request ? The extproc processing response is intercepted by extproc itself, maybe I misunderstood your proposal.

**Comment by mathetake** [comment]
My idea was to send the request to the Gateway's listener so the request will go through all the filters not only extproc (I'm not sure what you meant by extproc listener though)

**Comment by arkodg** [comment]
there's another attachment point available - ext_proc as an upstream filter in the xds cluster than can be used to perform backend specific request transformations

**Comment by mathetake** [comment]
Yeah that's exactly what I commented in https://github.com/envoyproxy/ai-gateway/issues/90#issuecomment-2743984765 but with extproc the entire body must go across twice which I believe is costly and why dynamic modules is really a good fit

**Comment by yuzisun** [comment]
> Yeah that's exactly what I commented in [#90 (comment)](https://github.com/envoyproxy/ai-gateway/issues/90#issuecomment-2743984765) but with extproc the entire body must go across twice which I believe is costly and why dynamic modules is really a good fit

Right I figured it needs to send all the way back to the gateway listener. I was also looking into the custom response filter redirection https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/http/custom_response/redirect_policy/v3/redirect_policy.proto#envoy-v3-api-msg-extensions-http-custom-response-redirect-policy-v3-redirectpolicy, this does the envoy internal redirection, would this work?

**Comment by yuzisun** [comment]
> there's another attachment point available - ext_proc as an upstream filter in the xds cluster than can be used to perform backend specific request transformations

@arkodg are you suggesting splitting the ext proc to downstream and upstream filter ? upstream filter does backend specific request transformation and upstream authentication, this will make fallback and rerouting easier. kgateway is doing that https://github.com/kgateway-dev/kgateway/blob/40fe7026aca5cb6ed38a3c8f1aafdff315774006/internal/kgateway/extensions2/plugins/backend/ai/ai_httpfilters.go#L104.

**Comment by mathetake** [comment]
> > there's another attachment point available - ext_proc as an upstream filter in the xds cluster than can be used to perform backend specific request transformations
> 
> [@arkodg](https://github.com/arkodg) are you suggesting splitting the ext proc to downstream and upstream filter ? upstream filter does backend specific request transformation and upstream authentication, this will make fallback and rerouting easier. kgateway is doing that https://github.com/kgateway-dev/kgateway/blob/40fe7026aca5cb6ed38a3c8f1aafdff315774006/internal/kgateway/extensions2/plugins/backend/ai/ai_httpfilters.go#L104.

@yuzisun yes, that's what arko meant. My comment above https://github.com/envoyproxy/ai-gateway/issues/34#issuecomment-2816779693 is a concern about using two extension points - one at the listener level which will decide upstream(backend) and the other is doing the all the backend specific work (transformation and authentication) since at both points we need to pass the entire body across envoy<>extproci (twice). In addition, the Envoy's native retry cannot change the already-chosen cluster, so we need something like https://github.com/envoyproxy/envoy/issues/38859 to achieve what we want.

**Comment by mathetake** [comment]
> > Yeah that's exactly what I commented in [#90 (comment)](https://github.com/envoyproxy/ai-gateway/issues/90#issuecomment-2743984765) but with extproc the entire body must go across twice which I believe is costly and why dynamic modules is really a good fit
> 
> Right I figured it needs to send all the way back to the gateway listener. I was also looking into the custom response filter redirection https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/http/custom_response/redirect_policy/v3/redirect_policy.proto#envoy-v3-api-msg-extensions-http-custom-response-redirect-policy-v3-redirectpolicy, this does the envoy internal redirection, would this work?

i am not sure if we can use custom response filter (which is also marked WIP in Envoy API) for this purpose since by the time we retry, the original request stream is at the response header phase (that's when we know the request failed) in extporc. We can simply make HTTP request from the extproc to directly gateway listener

**Comment by yuzisun** [comment]
> > > Yeah that's exactly what I commented in [#90 (comment)](https://github.com/envoyproxy/ai-gateway/issues/90#issuecomment-2743984765) but with extproc the entire body must go across twice which I believe is costly and why dynamic modules is really a good fit
> > 
> > 
> > Right I figured it needs to send all the way back to the gateway listener. I was also looking into the custom response filter redirection https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/http/custom_response/redirect_policy/v3/redirect_policy.proto#envoy-v3-api-msg-extensions-http-custom-response-redirect-policy-v3-redirectpolicy, this does the envoy internal redirection, would this work?
> 
> i am not sure if we can use custom response filter (which is also marked WIP in Envoy API) for this purpose since by the time we retry, the original request stream is at the response header phase (that's when we know the request failed) in extporc. We can simply make HTTP request from the extproc to directly gateway listener

ok, btw we need a request body cache store for the retries in the extproc as it is not directly available in the response time handlers.

**Comment by mathetake** [comment]
i think it's closed in v0.2 release!
